### PR TITLE
Pytorch 1.10.x

### PR DESCRIPTION
**Pull request by tobijk**
* On osx-64 build goes through, but packaging hangs for several hours, reason unknown. The only thing I see is that something is being run through multiprocessing.

* s390x cannot be built in prefect, the exact reason is unknown but the probable cause is resource exhaustion. Building manually outside of Docker is - interestingly - possible.

* The Windows build has to be performed manually, as well.

* On linux-64 all core tests pass on all other platforms we have failing tests
    - linux-64:
        https://gist.github.com/tobijk/db988150dc49295819f6b32a9edc6574
    - linux-aarch64:
        https://gist.github.com/tobijk/3af6b09c079f72afa3a432592175900d
    - linux-ppc64le:
        https://gist.github.com/tobijk/fff95ad9bc100b4b66a5d42e3d873914
    - linux-s390x:
        https://gist.github.com/tobijk/25c1a8fb693566088c901c88ca50d941
    - osx-64:
        https://gist.github.com/tobijk/29a11adc48fe426d6945213082be5f21
    - osx-arm64:
        https://gist.github.com/tobijk/56bd716e2ecfe1b350cb66a6eec0ab78

### CODE REVIEW

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -1,14 +1,12 @@
 # cuda 9.2, gcc 7.3, cudnn 7.6
 # cuda 10.0, gcc 7.3, cudnn 7.6
 # cuda 10.1, gcc 7.3, cudnn 7.6
-c_compiler_version:           # [linux]
+c_compiler_version:           # [unix]
   - 7.3                       # [linux]
-c_compiler_version:           # [osx]
-  - 4                        # [osx]
-cxx_compiler_version:         # [linux]
+  - 4                         # [osx]
```
Comment (katietz) [inline recipe/conda_build_config.yaml]: is it really still depending on clang 4 for osx?

Comment (katietz) [inline recipe/conda_build_config.yaml]: another thing here is that for osx-arm64 there is no clang 4 version, so this needs more specialization

Comment (katietz) [inline recipe/conda_build_config.yaml]: same here as above with c_compiler_version

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -1,14 +1,12 @@
...
+cxx_compiler_version:         # [unix]
   - 7.3                       # [linux]
-cxx_compiler_version:         # [osx]
-  - 4                        # [osx]
+  - 4                         # [osx]
 cross_compiler_target_platform:   # [win]
   - win-64                        # [win]
```
Comment (katietz) [inline recipe/conda_build_config.yaml]: is this required? I suspect it.  at least it looks pretty strange to describe here cross for win-64

File: recipe/conda_build_config.yaml [hunk 3]
```diff
@@ -18,20 +16,21 @@ cudatoolkit:
...
+  - openblas    # [not x86]
 mkl:
-  - 2019
+  - 2020        # [linux]
+  - 2019        # [osx]
+openblas:
+  - 0.3.10
 numpy:
-  - 1.11
+  - 1.16
```
Comment (katietz) [inline recipe/conda_build_config.yaml]: this looks wrong.  please see how python/numpy are declared in aggregate's cbc.  using here 1.16 in general is for sure not ok, and might be also one issue why python 3.10 doesn't here ... as for python 3.10 numpy 1.21 is mandatory

File: recipe/conda_build_config.yaml [hunk 4]
```diff
@@ -18,20 +16,21 @@ cudatoolkit:
...
 MACOSX_DEPLOYMENT_TARGET:
   - 10.10
```
Comment (katietz) [inline recipe/conda_build_config.yaml]: this needs adjustment to distinguish between osx-64 and osx-arm64.  I assume this block should be made used only for osx-64

**Review by katietz — CHANGES_REQUESTED** [review]

File: recipe/conda_build_config.yaml (hunk 1 above)
Comment (tobijk) [reply recipe/conda_build_config.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

Comment (tobijk) [reply recipe/conda_build_config.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

Comment (tobijk) [reply recipe/conda_build_config.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

File: recipe/conda_build_config.yaml (hunk 2 above)
Comment (tobijk) [reply recipe/conda_build_config.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

File: recipe/conda_build_config.yaml (hunk 3 above)
Comment (tobijk) [reply recipe/conda_build_config.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

File: recipe/conda_build_config.yaml (hunk 4 above)
Comment (tobijk) [reply recipe/conda_build_config.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -41,24 +46,27 @@ requirements:
     - magma                           # [pytorch_variant == "gpu"]
     # other requirements
     - python
-    - numpy
+    - numpy-devel {{ numpy }}
```
[the remark this answers is not in the document]
Comment (tobijk) [reply recipe/meta.yaml]: This comment is based on an older version of the PR and is not applicable anymore.

File: recipe/build.sh [hunk 6]
```diff
@@ -1,51 +1,94 @@
...
+
+# This must be unset, else PyTorch complains.
+unset CMAKE_INSTALL_PREFIX
+
 export PYTORCH_BUILD_VERSION=$PKG_VERSION
 export PYTORCH_BUILD_NUMBER=$PKG_BUILDNUM
 
-export USE_NINJA=OFF
+export TH_BINARY_BUILD=1
+export USE_NINJA=0
```
[the remark this answers is not in the document]
Comment (tobijk) [reply recipe/build.sh:49]: To be honest, I just left it the way it was. We can try with Ninja.

File: recipe/bld.bat [hunk 7]
```diff
@@ -12,35 +17,53 @@ if "%pytorch_variant%" == "gpu" (
...
-
-set CMAKE_INCLUDE_PATH=%LIBRARY_PREFIX%\include
-set LIB=%LIBRARY_PREFIX%\lib;%LIB%
+if "%desired_cuda%" == "9.0" (
+    set TORCH_CUDA_ARCH_LIST=%TORCH_CUDA_ARCH_LIST%;6.0;7.0
+)
+if "%desired_cuda%" == "9.2" (
+    set TORCH_CUDA_ARCH_LIST=%TORCH_CUDA_ARCH_LIST%;6.0;6.1;7.0
+)
+if "%desired_cuda%" == "10.0" (
```
[the remark this answers is not in the document]
Comment (tobijk) [reply recipe/bld.bat:33]: I'll note this for when we update the GPU version.

File: recipe/meta.yaml [hunk 8]
```diff
@@ -10,93 +10,171 @@ source:
...
+    - openblas                        # [blas_impl == "openblas"]
+    # OpenMP
+    - intel-openmp                    # [(osx or win) and x86]
+    - llvm-openmp                     # [osx and arm64]
+    # Other requirements
+    - cffi
+    - future
+    - libuv                           # [win]
+    # The test suite uses features only available in numpy 1.17+.
+    - numpy >=1.17
```
[the remark this answers is not in the document]
Comment (tobijk) [reply recipe/meta.yaml]: Done

**Review by katietz — APPROVED** [review]
thanks for adjustments.  Gpu part post-spawned until we actual tackle GPU

**Pull request merged**